File: carsharing_server/src/main/java/org/example/carsharing_server/Car/Car.java
package org.example.carsharing_server.Car;


import com.fasterxml.jackson.annotation.JsonIgnore;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.datatype.hibernate6.Hibernate6Module;
import jakarta.persistence.*;
import jakarta.validation.constraints.Min;
import jakarta.validation.constraints.NotBlank;
import jakarta.validation.constraints.Pattern;
import org.example.carsharing_server.Booking.Booking;
import org.example.carsharing_server.CarMaintenance.CarMaintenance;
import org.example.carsharing_server.Location.Location;
import org.example.carsharing_server.User.User;

import java.util.ArrayList;
import java.util.List;

@Entity
@Table
public class Car {

    @Id
    @Column(unique = true)
    @Pattern(regexp = "[a-zA-Z]..[0-9]..")
    @NotBlank
    private String licensePlate;

    @NotBlank
    private String model;

    @Min(1672)
    private int releaseYear;

    @Min(1)
    private double price;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "locationID", referencedColumnName = "locationID")
    private Location location;

    private boolean availabilityStatus=true;

    @OneToMany(mappedBy = "car", cascade = CascadeType.ALL)
    @JsonIgnore
    private List<CarMaintenance> maintenanceRecord;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "userID", referencedColumnName = "userID")
    private User owner;

    @OneToMany(mappedBy = "car", cascade = CascadeType.ALL)
    @JsonIgnore
    private List<Booking> bookings;


    public Car() {

    }

    public Car(String plate){
        this.licensePlate = plate;
    }

    public Car(String model, int releaseYear, String licensePlate, Location location, User owner, double price) {
        this.model = model;
        this.releaseYear = releaseYear;
        this.licensePlate = licensePlate;
        this.price = price;
        this.owner = owner;
        this.location = location;
        this.maintenanceRecord = new ArrayList<CarMaintenance>();
        this.availabilityStatus = true;
        this.bookings = new ArrayList<>();
    }


    public String getLicensePlate() {
        return licensePlate;
    }

    public void setLicensePlate(String licensePlate) {
        this.licensePlate = licensePlate;
    }

    public String getModel() {
        return model;
    }

    public void setModel(String model) {
        this.model = model;
    }

    public int getReleaseYear() {
        return releaseYear;
    }

    public void setReleaseYear(int releaseYear) {
        this.releaseYear = releaseYear;
    }

    public Location getLocation() {
        return location;
    }

    public void setLocation(Location location) {
        this.location = location;
    }

    public boolean getAvailabilityStatus() {
        return availabilityStatus;
    }

    public void setAvailabilityStatus(boolean availabilityStatus) {
        this.availabilityStatus = availabilityStatus;
    }

    public List<CarMaintenance> getMaintenanceRecord() {
        return maintenanceRecord;
    }

    public void addMaintenance(CarMaintenance maintenance) {
        maintenanceRecord.add(maintenance);
    }

    public List<Booking> getBookings() {
        return bookings;
    }

    public void setBookings(List<Booking> bookings) {
        this.bookings = bookings;
    }

    public User getOwner() {
        return owner;
    }

    public void setOwner(User owner) {
        this.owner = owner;
    }

    public double getPrice() {
        return price;
    }

    public void setPrice(int price) {
        this.price = price;
    }

    @Override
    public String toString() {
        ObjectMapper objectMapper = new ObjectMapper();
        objectMapper.registerModule(new Hibernate6Module()
                .enable(Hibernate6Module.Feature.SERIALIZE_IDENTIFIER_FOR_LAZY_NOT_LOADED_OBJECTS)
                .enable(Hibernate6Module.Feature.WRITE_MISSING_ENTITIES_AS_NULL));
        try {
            return objectMapper.writeValueAsString(this);
        } catch (Exception e) {
            throw new RuntimeException("Object not mappable");
        }
    }

}

File: carsharing_client/Carshare/car.cpp
#include "car.h"

Car::Car(const QString &name, const QString &model, double price,
         const QString &location, const QString &owner, const QString &licensePlate, int releaseYear)
    : m_name(name), m_model(model), m_price(price),
    m_location(location), m_owner(owner), m_licensePlate(licensePlate), m_releaseYear(releaseYear) {}

QString Car::name() const {
    return m_name;
}

void Car::setName(const QString &name) {
    m_name = name;
}

QString Car::model() const {
    return m_model;
}

void Car::setModel(const QString &model) {
    m_model = model;
}

double Car::pricePerDay() const {
    return m_price;
}

void Car::setPricePerDay(double pricePerDay) {
    m_price = pricePerDay;
}

QString Car::location() const {
    return m_location;
}

void Car::setLocation(const QString &location) {
    m_location = location;
}

QString Car::owner() const {
    return m_owner;
}

void Car::setOwner(const QString &owner) {
    m_owner = owner;
}

QString Car::licensePlate() const {
    return m_licensePlate;
}

void Car::setLicensePlate(const QString &licensePlate) {
    m_licensePlate = licensePlate;
}

int Car::releaseYear() const {
    return m_releaseYear;
}

void Car::setReleaseYear(int releaseYear) {
    m_releaseYear = releaseYear;
}
